### PR TITLE
[tailscale] net: add enforcement hooks

### PR DESCRIPTION
**Pull request by bradfitz**
Updates tailscale/go#55
Updates tailscale/corp#8944

### CODE REVIEW

File: src/net/dial.go [hunk 1]
```diff
@@ -209,6 +209,24 @@ func parseNetwork(ctx context.Context, network string, needsProto bool) (afnet s
 	return "", 0, UnknownNetworkError(network)
 }
 
+// SetResolveEnforcer set a program-global resolver enforcer that can cause resolvers to
+// fail based on the context and/or other arguments.
+//
+// f must be non-nil, it can only be called once, and must not be called
```
Comment (danderson) [inline src/net/dial.go:215]: Do these constraints add up to "must be called from init()" ?

**Review by danderson — APPROVED** [review]

File: src/net/dial.go [hunk 2]
```diff
@@ -209,6 +209,24 @@ func parseNetwork(ctx context.Context, network string, needsProto bool) (afnet s
...
+// concurrent with any dial/resolve.
+func SetResolveEnforcer(f func(ctx context.Context, op, network, addr string, hint Addr) error) {
+	if f == nil {
+		panic("nil func")
+	}
+	if dialEnforcer != nil {
```
Comment (maisem) [inline src/net/dial.go]: should this be `resolveEnforcer`?

File: src/net/dial.go [hunk 3]
```diff
@@ -209,6 +209,24 @@ func parseNetwork(ctx context.Context, network string, needsProto bool) (afnet s
...
+		panic("already called")
+	}
+	resolveEnforcer = f
+}
+
+// dialEnforce, if non-nil
```
Comment (maisem) [inline src/net/dial.go]: same

File: src/net/dial.go (hunk 1 above)
Comment (bradfitz) [reply src/net/dial.go:215]: Not necessarily. Early in main is fine too.

**Review by maisem — APPROVED** [review]

**Pull request merged**